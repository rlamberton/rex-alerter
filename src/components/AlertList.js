import { useState } from 'react';
import Alert from './Alert';
import getNewAlerts from '../service/alertService';

const REFRESH_PERIOD = 5000;    // Every 5 seconds
var listOfAlerts, updateAlerts;

/**
 * React functional component for the Alert List
 * 
 * Component state holds the list of alerts
 */
function AlertList(props) {
    [listOfAlerts, updateAlerts] = useState([]);

    return  <div>
                <ul className='alertList'>
                    {listOfAlerts.map((item, index) =>
                        <Alert key={index} item={item}/>
                    )}
                </ul>
            </div>;
};

/**
 * Timer to fire every 5 seconds, to fetch the latest alerts from the API
 */
setInterval(() => {
    const newAlertsPromise = getNewAlerts();
    newAlertsPromise.then((newAlerts) => {
        if (newAlerts && newAlerts.length) {
            // Update the state with the new list
            updateAlerts(oldAlerts => [...oldAlerts].concat(newAlerts));

            // Auto-scroll to new element
            const alertListElement = document.querySelector(".alertList");
            let lastAlert = alertListElement.lastElementChild;
            if (lastAlert) {
                lastAlert.scrollIntoViewIfNeeded(true);
            }
        }
    });

}, REFRESH_PERIOD);

export default AlertList;
