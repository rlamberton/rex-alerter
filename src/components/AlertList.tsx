import { useState, useEffect } from 'react';
import Alert from './Alert';
import AlertType from '../types/AlertType';
import LastUpdated from './LastUpdated';
import getNewAlerts from '../service/AlertService';

const REFRESH_PERIOD : number = 5000;    // Every 5 seconds

/**
 * React functional component for the Alert List
 * 
 * Component state holds the list of alerts
 */
const alertList : React.FC = function AlertList(props) {
    const [listOfAlerts, updateAlerts] = useState<AlertType[]>([]);
    const [lastUpdatedTime, setLastUpdatedTime] = useState<string>(new Date().toLocaleTimeString());

    /* Timer to fire every 5 seconds, to fetch the latest alerts from the API */
    useEffect(() => {
        const interval = setInterval(() => {
            const newAlertsPromise = getNewAlerts();
            newAlertsPromise.then((newAlerts) => {
                if (newAlerts && newAlerts.length) {
                    // Update the state with the new list
                    updateAlerts(oldAlerts => [...oldAlerts].concat(newAlerts));
        
                    // Auto-scroll to end of page if required
                    document.querySelector("#lastUpdated")?.scrollIntoView(true);
                }
            });
        
            setLastUpdatedTime(new Date().toLocaleTimeString());
        
        }, REFRESH_PERIOD);

        return () => clearInterval(interval);
    }, []);

    return  <div>
                <ul className='alertList'>
                    {listOfAlerts.map((item, index) =>
                        <Alert key={index} item={item}/>
                    )}
                </ul>
                <LastUpdated time={lastUpdatedTime}/>
            </div>;
};

export default alertList;
