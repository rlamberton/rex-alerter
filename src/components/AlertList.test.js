import { render } from '@testing-library/react';
import { AlertList, fetchAlerts } from './AlertList';

const data = {
    url: 'http://url.com',
    time: '10:30',
    logoUrl: 'http:/logo.com',
    symbol: 'XYZ',
    name: 'crypto-name',
    volume: 1000,
    oldPrice: 2000,
    newPrice: 2200,
    pctChange : 10
};

const MOCK_RESULTS = {
    ETH: {
        symbol: 'ETH-BTC',
        status: 'OFFLINE'
    },
    'ETH-BTC': {
        symbol: '',
        status: 'ONLINE'
    }
}

function wait(ms){
    var start = new Date().getTime();
    var end = start;
    while(end < start + ms) {
      end = new Date().getTime();
   }
}

test('check snapshot', () => {
    const { asFragment } = render(<AlertList item={data} />);
    expect( asFragment() ).toMatchSnapshot();
});

test('call fetchAlerts', () => {
    jest.spyOn(global, 'fetch').mockResolvedValue({
        ok: true,
        json: jest.fn().mockResolvedValue(
            { eth: { symbol: 'ETH-BTC', quoteVolume: 20, askRate: 100, name: "Etherium" } ,
            'ETH-BTC': { symbol: 'ETH', status: 'ONLINE', name: "Etherium" },
            ETH: { name: 'Etherium', symbol: 'XYZ', status: 'ONLINE'}}
        )
    })

    const mock1 = jest.fn();
    const mock2 = jest.fn();
    const result = fetchAlerts(mock1, mock2);

    jest.restoreAllMocks();

    expect(mock1).toBeCalledTimes(0);
    expect(mock2).toBeCalledTimes(1);
});
