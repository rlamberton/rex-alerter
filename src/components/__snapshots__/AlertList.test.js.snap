// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check snapshot 1`] = `
<DocumentFragment>
  <div>
    <ul
      class="alertList"
      data-testid="alert-list"
    />
    <div
      data-testid="alert-list-last-updated"
      id="lastUpdated"
    >
      Last updated at 
    </div>
  </div>
</DocumentFragment>
`;
